src/serializer.ts

```diff
@@ -0,0 +1,115 @@
+    */
+    'kind'?: string;
+    'metadata'?: V1ObjectMeta;
+
+    static attributeTypeMap: AttributeType[] = [
+        {
+            "name": "apiVersion",
+            "baseName": "apiVersion",
+            "type": "string"
+        },
+        {
+            "name": "kind",
+            "baseName": "kind",
+            "type": "string"
+        },
+        {
+            "name": "metadata",
+            "baseName": "metadata",
+            "type": "V1ObjectMeta"
+        },
+    ];
+}
+
+const isKubernetesObject = (data: unknown): boolean =>
+    !!data && typeof data === "object" && 'apiVersion' in data && 'kind' in data;
+
+/**
+ * Wraps the ObjectSerializer to support custom resources and generic Kubernetes objects.
+ */
+class KubernetesObjectSerializer {
+
+    private static _instance: KubernetesObjectSerializer;
+
+    public static get instance() {
+        if (this._instance) {
+            return this._instance;
+        }
+        this._instance = new KubernetesObjectSerializer();
+        return this._instance;
+    }
+
+    private constructor() {}
+
+    public serialize(data: any, type: string) {
+        const obj = ObjectSerializer.serialize(data, type);
+        if(obj !== data) {
+            return obj;
+        }
+
+        if (!isKubernetesObject(data)) {
+            return obj;
+        }
+
+        const instance: Record<string, any> = {};
+        const attributeTypes = KubernetesObject.attributeTypeMap;
+        for (let index = 0; index < attributeTypes.length; index++) {
+            let attributeType = attributeTypes[index];
+            instance[attributeType.name] = ObjectSerializer.serialize(data[attributeType.baseName], attributeType.type);
+        }
+        // add all unknown properties as is.
+        for (const [key, value] of Object.entries(data)) {
+            if (attributeTypes.find((t) => t.name === key)) {
+                continue;
+            }
+            instance[key] = value;
+        }
+        return instance;
+
+    }
+
+    public deserialize(data: any, type: string) {
+        const obj = ObjectSerializer.deserialize(data, type);
+        if (obj !== data) {
+            // the serializer knows the type and already deserialized it.
+            return obj;
+        }
+
+        if (!isKubernetesObject(data)) {
+            return obj;
+        }
+
+        const instance = new KubernetesObject();
+        const attributeTypes = KubernetesObject.attributeTypeMap;
+        for (let index = 0; index < attributeTypes.length; index++) {
+            let attributeType = attributeTypes[index];
+            instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type);
+        }
```

Comment:
Similar here, or is there a reason for not using `for ... in`?

Reply:
no reason. And I agree that the `for in` / `for of` loops are more readable.
Also changed it here